File: contact-route-planner/backend/geocode.py
from geopy.geocoders import Nominatim

geolocator = Nominatim(user_agent="contact_route_planner")

def geocode_address(street, city, state, zip_code):
    full_address = f"{street}, {city}, {state} {zip_code}"

    location = geolocator.geocode(full_address)

    if location:
        return location.latitude, location.longitude

    return None, None

File: backend-mern/utils/geocode.js
import NodeGeocoder from "node-geocoder";

const options = {
  provider: "openstreetmap"
};

const geocoder = NodeGeocoder(options);

export async function geocodeAddress(
  street,
  city,
  state,
  zip
) {
  try {
    const address = `${street}, ${city}, ${state} ${zip}`;

    const results = await geocoder.geocode(address);

    if (results.length === 0) {
      return {
        latitude: null,
        longitude: null
      };
    }

    return {
      latitude: results[0].latitude,
      longitude: results[0].longitude
    };
  } catch (error) {
    console.error("Geocoding error:", error);

    return {
      latitude: null,
      longitude: null
    };
  }
}